_posts/2023-10-11-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,485 @@
+```shell
+kubectl get kafka <kafka_cluster_name> -o=jsonpath='{.status.listeners[*].bootstrapServers}{"\n"}'
+```
+
+This command retrieves the bootstrap addresses exposed by listeners for client connections on a Kafka cluster. 
+Listeners define how clients communicate with the Kafka brokers, ensuring that you connect to the correct advertised listener port.
+In production environments, you can use a single load balancer, or a list of brokers in place of a single bootstrap server.
+
+**GROUP_ID**
+
+The consumer group identifier.
+
+**POLL_TIMEOUT_MS** 
+
+The maximum time to wait for new messages during each poll.
+
+**TOPIC_NAME**
+
+The name of the topic where the consumer application receives its messages.
+
+**NUM_MESSAGES**
+      
+The number of messages the client consumes before it shuts down.
+
+**PROCESSING_DELAY_MS**
+
+This constant can simulate typical message consumption patterns, which is especially useful for testing.
+In Kafka, messages typically capture streams of events, so introducing delays can help simulate peak or average event rates.
+
+These constants give us some control over the consumer application's behavior.
+We can use the `NUM_MESSAGES` and `PROCESSING_DELAY_MS` values to adjust the message receiving rate.
+With the current settings, a total of 50 messages are received with a delay of 1 second between them.
+
+### Example consumer application
+
+Time to create our client.
+We want our example client to operate as follows:
+
+* Create a Kafka consumer instance, which is responsible for receiving messages from a Kafka topic.
+* Subscribe to a specific Kafka topic (in this case, `my-topic`) to start receiving messages from it.
+* Handle partition assignment, revocation, and loss.
+* Use deserializer classes that handle the transformation of message keys and values into a format suitable for processing. 
+* Consume messages until a specified number of messages (`NUM_MESSAGES`) is reached.
+* Control the rate at which messages are consumed by introducing delays between each message using our `PROCESSING_DELAY_MS` value.
+* Commit the message offsets asynchronously to Kafka to record the progress of message consumption.
+* Handle errors that may occur during message consumption, determining when a message should be retried and when an error is considered non-recoverable. 
+
+**consumer configuration**
+
+We'll specify the minimal configuration properties required for a consumer instance:
+
+* `BOOTSTRAP_SERVERS_CONFIG` for connection to the Kafka brokers. This picks up the value of the `BOOTSTRAP_SERVERS` constant.
+* `CLIENT_ID_CONFIG` that uses a randomly generated UUID as a client ID for tracking the source of requests.
+* `ENABLE_AUTO_COMMIT_CONFIG` to disable automatic offset committing by Kafka. We will control offset commits in our application.
+* `AUTO_OFFSET_RESET_CONFIG` to define the starting point for consuming messages when no committed offset is found for the configured group ID and assigned partition.  
```

Comment:
This does not seem configurable in the code below, it's hard coded to "earliest"